README
NAME
    Template::Plugin::IPAddr - Template::Toolkit plugin handling
    IP-addresses

SYNOPSIS
      # Create IPAddr object via USE directive...
      [% USE IPAddr %]
      [% USE IPAddr(prefix) %]

      # ...or via new
      [% ip = IPAddr.new(prefix) %]

      # Methods that return the different parts of the prefix
      [% IPAddr.addr %]
      [% IPAddr.addr_cidr %]
      [% IPAddr.cidr %]
      [% IPAddr.network %]
      [% IPAddr.netmask %]
      [% IPAddr.wildcard %]

      # Methods for retrieving usable IP-adresses from a prefix
      [% IPAddr.first %]
      [% IPAddr.last %]

DESCRIPTION
    This module implements an "IPAddr" class for handling IPv4 and
    IPv6-address in an object-orientated way. The module is based on
    NetAddr::IP and works on IPv4 as well as IPv6-addresses.

    You can create a "IPAddr" object via the "USE" directive, adding any
    initial prefix as an argument.

      [% USE IPAddr %]
      [% USE IPAddr(prefix) %]

    Once you've got a "IPAddr" object, you can use it as a prototype to
    create other "IPAddr" objects with the new() method.

      [% USE IPAddr %]
      [% ip = IPAddr.new(prefix) %]

    After creating an "IPaddr" object, you can use the supplied methods for
    retrieving properties of the prefix.

      [% USE IPAddr('10.0.0.0/24') %]
      [% IPAddr.netmask %]   # 255.255.255.0
      [% IPAddr.first %]     # 10.0.0.1
      [% IPAddr.last %]      # 10.0.0.254

METHODS
  new
    Creates a new IPAddr object using an initial value passed as a
    positional parameter. Any string which is accepted by NetAddr::IP->new
    can be used as a parameter.

      [% USE IPAddr %]
      [% USE IPAddr(prefix) %]
      [% ip = IPAddr.new(prefix) %]

    Examples of (recommended) formats of initial parameters that can be
    used:

      # IPv4
      n.n.n.n             # Host address
      n.n.n.n/m           # CIDR notation
      n.n.n.n/m.m.m.m     # address + netmask

      # IPv6
      x:x:x:x:x:x:x:x     # Host address
      x:x:x:x:x:x:x:x/m   # CIDR notation
      ::n.n.n.n           # IPv4-compatible IPv6 address

    When used as "[% USE IPAddr %]" the prefix assigned internally is
    "0.0.0.0/0"

  addr
    Returns the address part of the prefix as written in the initial value.

      [% USE IPAddr('10.1.1.1/24') %]
      [% IPAddr.addr %]  # 10.1.1.1

      [% USE IPAddr('2001:DB8::DEAD:BEEF') %]
      [% IPAddr.addr %]  # 2001:db8::dead:beef

  addr_cidr
    Returns the *address* in CIDR notation, i.e. as "address/prefixlen".

      [% USE IPAddr('10.1.1.1/255.255.255.0') %]
      [% IPAddr.addr_cidr %]   # 10.1.1.1/24

      [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
      [% IPAddr.addr_cidr %]  # 2001:db8:a:b:c:d:e:f/48

  cidr
    Returns the *prefix* in CIDR notation, i.e. as "network/prefixlen".

      [% USE IPAddr('10.1.1.1/255.255.255.0') %]
      [% IPAddr.cidr %]   # 10.1.1.0/24

      [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
      [% IPAddr.cidr %]  # 2001:db8:a::/48

    Note that differs from the "cidr" method in NetAddr::IP (which returns
    "address/prefixlen"). You can retrieve an address on that format by
    using the "addr_cidr" method.

  first
    Returns the first usable IP-address within the prefix.

      [% USE IPAddr('10.0.0.0/16') %]
      [% IPAddr.first %]   # 10.0.0.1

  last
    Returns the last usable IP-address within the prefix.

      [% USE IPAddr('10.0.0.0/16') %]
      [% IPAddr.last %]   # 10.0.255.254

  network
    Returns the network part of the prefix.

      [% USE IPAddr('10.1.1.1/24') %]
      [% IPAddr.network %]   # 10.1.1.0

      [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
      [% IPAddr.network %]  # 2001:db8:a::

  netmask
    Returns the netmask part of the prefix.

      [% USE IPAddr('10.1.1.1/24') %]
      [% IPAddr.netmask %]   # 255.255.255.0

  wildcard
    Returns the netmask of the prefix in wildcard format (the netmask with
    all bits inverted).

      [% USE IPAddr('10.1.1.1/24') %]
      [% IPAddr.wildcard %]   # 0.0.0.255

NOTES
    Please note the subtle, but important, difference between "addr_cidr"
    and "cidr" (see "cidr" for an explanation).

    Not all methods are applicable in a IPv6 context. For example there are
    no notation of netmask or wildcard in IPv6, and the first and last
    returns values of no use.

    When using IPv6 mapped IPv4 addresses, the "dot notation" is lost in the
    process. For example:

      [% USE IPAddr('::192.0.2.1') %]

    then

      [% IPAddr.addr %]

    will print "::c000:201".

SEE ALSO
    Template, "PLUGINS" in Template::Manual::Config, NetAddr::IP

SOURCE CODE REPOSITORY
    The source code for this module is held in a public git repository at
    <https://github.com/hemmop/template-plugin-ipaddr.git>

AUTHOR
    Per Carlson "pelle@cpan.org"

COPYRIGHT AND LICENSE
    Copyright 2013,2014 Per Carlson

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl 5.10.0.

    For more details, see the full text of the licenses at
    <http://opensource.org/licenses/Artistic-2.0>, and
    <http://opensource.org/licenses/GPL-2.0>.



Changes
{{$NEXT}}
	- Added method addr_cidr
	- Use Dist::Zilla for packaging

0.02	2013-09-10
	- Added Changes
	- Updated Makefile.PL

0.01	2013-09-07
	- First release


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;

my %args = (
  NAME           => 'Template::Plugin::IPAddr',
  AUTHOR         => 'Per Carlson <pelle@cpan.org>',
  VERSION_FROM   => 'lib/Template/Plugin/IPAddr.pm',
  ABSTRACT_FROM  => 'lib/Template/Plugin/IPAddr.pm',
  LICENSE        => 'perl',
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {
    'Test::More'          => 0,
    'Test::Pod::Coverage' => 0,
  },
  PREREQ_PM => {
    'NetAddr::IP' => '4.062',
    'Template'    => 0,
  },
  META_MERGE => {
    resources => {
      repository => 'https://github.com/hemmop/template-plugin-ipaddr',
      bugtracker => 'https://github.com/hemmop/template-plugin-ipaddr/issues',
    },
  },
);

unless (eval { ExtUtils::MakeMaker->VERSION(6.63_03) }) {
  my $tr = delete $args{TEST_REQUIRES};
  my $br = $args{BUILD_REQUIRES};
  for my $mod (keys %$tr) {
    if (exists $br->{$mod}) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.56) }) {
  my $br = delete $args{BUILD_REQUIRES};
  my $pp = $args{PREREQ_PM};
  for my $mod (keys %$br) {
    if (exists $pp->{$mod}) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.45_01) }) {
  delete $args{META_MERGE};
}

WriteMakefile(%args);



lib/Template/Plugin/IPAddr.pm
package Template::Plugin::IPAddr;
# ABSTRACT: Template::Toolkit plugin handling IP-addresses

use strict;
use warnings;
use base 'Template::Plugin';

use NetAddr::IP qw{ :lower };
use Scalar::Util qw{ blessed };

use overload '""' => sub { shift->cidr };

sub new {
  my ($class, $context, $arg) = @_;
  # When used as [% USE IPAddr %] or [% USE IPAddr(addr) %]
  # $context is a Template::Context object, and $arg is filled
  # with the arguments to IPAddr (undef resp. addr here).
  # When used as [% ip = NetAddr.new(addr) %], $context contain
  # the addr.
  my $addr = blessed $context ? $arg : $context;
  return bless { _cidr => NetAddr::IP->new($addr) }, ref $class || $class;
}

sub addr { return _compact(shift->{_cidr}) }

sub addr_cidr {
  my $self = shift;
  return $self->addr . '/' . $self->{_cidr}->masklen;
}

sub cidr {
  my $self = shift;

  # we can't use the cidr method because we want network/prefix,
  # and cidr returns addr/prefix.
  #
  # return an ipv6 address in compact format (with '::').
  return $self->network . '/' . $self->{_cidr}->masklen;
}

sub first    { return _compact(shift->{_cidr}->first) }
sub last     { return _compact(shift->{_cidr}->last) }
sub netmask  { return shift->{_cidr}->mask }
sub network  { return _compact(shift->{_cidr}->network) }
sub wildcard { return scalar shift->{_cidr}->wildcard }

# This sub takes an NetAddr::IP object and returns
# the address in short notation if IPv6, or the
# address as is if IPv4.
sub _compact {
  my $ip = shift;
  return $ip->addr =~ /:/ ? $ip->short : $ip->addr;
}

1;

__END__

=pod

=head1 SYNOPSIS

  # Create IPAddr object via USE directive...
  [% USE IPAddr %]
  [% USE IPAddr(prefix) %]

  # ...or via new
  [% ip = IPAddr.new(prefix) %]

  # Methods that return the different parts of the prefix
  [% IPAddr.addr %]
  [% IPAddr.addr_cidr %]
  [% IPAddr.cidr %]
  [% IPAddr.network %]
  [% IPAddr.netmask %]
  [% IPAddr.wildcard %]

  # Methods for retrieving usable IP-adresses from a prefix
  [% IPAddr.first %]
  [% IPAddr.last %]

=head1 DESCRIPTION

This module implements an C<IPAddr> class for handling IPv4 and IPv6-address
in an object-orientated way.
The module is based on L<NetAddr::IP> and works on IPv4 as well as
IPv6-addresses.

You can create a C<IPAddr> object via the C<USE> directive, adding any initial
prefix as an argument.

  [% USE IPAddr %]
  [% USE IPAddr(prefix) %]

Once you've got a C<IPAddr> object, you can use it as a prototype to create
other C<IPAddr> objects with the new() method.

  [% USE IPAddr %]
  [% ip = IPAddr.new(prefix) %]

After creating an C<IPaddr> object, you can use the supplied methods for
retrieving properties of the prefix.

  [% USE IPAddr('10.0.0.0/24') %]
  [% IPAddr.netmask %]   # 255.255.255.0
  [% IPAddr.first %]     # 10.0.0.1
  [% IPAddr.last %]      # 10.0.0.254


=method new

Creates a new IPAddr object using an initial value passed as a positional
parameter. Any string which is accepted by L<< NetAddr::IP->new >> can be
used as a parameter.

  [% USE IPAddr %]
  [% USE IPAddr(prefix) %]
  [% ip = IPAddr.new(prefix) %]

Examples of (recommended) formats of initial parameters that can be used:

  # IPv4
  n.n.n.n             # Host address
  n.n.n.n/m           # CIDR notation
  n.n.n.n/m.m.m.m     # address + netmask

  # IPv6
  x:x:x:x:x:x:x:x     # Host address
  x:x:x:x:x:x:x:x/m   # CIDR notation
  ::n.n.n.n           # IPv4-compatible IPv6 address

When used as C<[% USE IPAddr %]> the prefix assigned internally is C<0.0.0.0/0>

=method addr

Returns the address part of the prefix as written in the initial value.

  [% USE IPAddr('10.1.1.1/24') %]
  [% IPAddr.addr %]  # 10.1.1.1

  [% USE IPAddr('2001:DB8::DEAD:BEEF') %]
  [% IPAddr.addr %]  # 2001:db8::dead:beef

=method addr_cidr

Returns the I<address> in CIDR notation, i.e. as C<address/prefixlen>.

  [% USE IPAddr('10.1.1.1/255.255.255.0') %]
  [% IPAddr.addr_cidr %]   # 10.1.1.1/24

  [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
  [% IPAddr.addr_cidr %]  # 2001:db8:a:b:c:d:e:f/48

=method cidr

Returns the I<prefix> in CIDR notation, i.e. as C<network/prefixlen>.

  [% USE IPAddr('10.1.1.1/255.255.255.0') %]
  [% IPAddr.cidr %]   # 10.1.1.0/24

  [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
  [% IPAddr.cidr %]  # 2001:db8:a::/48

Note that differs from the C<cidr> method in L<NetAddr::IP> (which
returns C<address/prefixlen>). You can retrieve an address on that
format by using the L</addr_cidr> method.

=method first

Returns the first usable IP-address within the prefix.

  [% USE IPAddr('10.0.0.0/16') %]
  [% IPAddr.first %]   # 10.0.0.1

=method last

Returns the last usable IP-address within the prefix.

  [% USE IPAddr('10.0.0.0/16') %]
  [% IPAddr.last %]   # 10.0.255.254

=method network

Returns the network part of the prefix.

  [% USE IPAddr('10.1.1.1/24') %]
  [% IPAddr.network %]   # 10.1.1.0

  [% USE IPAddr('2001:db8:a:b:c:d:e:f/48') %]
  [% IPAddr.network %]  # 2001:db8:a::

=method netmask

Returns the netmask part of the prefix.

  [% USE IPAddr('10.1.1.1/24') %]
  [% IPAddr.netmask %]   # 255.255.255.0

=method wildcard

Returns the netmask of the prefix in wildcard format (the netmask
with all bits inverted).

  [% USE IPAddr('10.1.1.1/24') %]
  [% IPAddr.wildcard %]   # 0.0.0.255

=head1 NOTES

Please note the subtle, but important, difference between C<addr_cidr>
and C<cidr> (see L</cidr> for an explanation).

Not all methods are applicable in a IPv6 context. For example there
are no notation of L<netmask> or L<wildcard> in IPv6, and the L<first>
and L<last> returns values of no use.

When using IPv6 mapped IPv4 addresses, the "dot notation" is lost
in the process. For example:

  [% USE IPAddr('::192.0.2.1') %]

then

  [% IPAddr.addr %]

will print C<::c000:201>.

=head1 SEE ALSO

L<Template>,
L<Template::Manual::Config/PLUGINS>,
L<NetAddr::IP>

=cut
